guides/graphql-subscriptions.md

```diff
@@ -0,0 +1,118 @@
+    id
+    currentTestingType
+  }
+}
+```
+
+A subscription is used when you want to receive information about data that has changed scoped to an individual event:
+
+```graphql
+subscription MyAppSubscription {
+  projectUpdated { # Project type
+    id
+    isLoadingConfig
+    isLoadingSetupNodeEvents
+  }
+}
+```
+
+The distinction between subscriptions & queries is that subscriptions can only have a single top-level field. You can think of the subscription like an "event name" in socket.io, except that it's strongly typed & part of the overall GraphQL schema, meaning that you don't need to do any extra work for it to merge into the normalized cache & refresh affected parts of the view. 
+
+```ts
+// will merge & update any views that depend on `isLoadingConfig` / `isLoadingSetupNodeEvents`
+useSubscription({ query: MyAppSubscriptionDocument })
+```
+
+We can also use subscriptions more granularly as an event emitter, basically as a strongly typed socket.io emitter:
+
+```graphql
+subscription OnSpecChange {
+  onSpecUpdate {
+    specPath
+    reason
+  }
+}
+```
+
+```ts
+useSubscription({ query: OnSpecChangeDocument }, (prev, next) => {
+  if (data.specPath === currentSpecPath && reason === 'DELETED') {
+    // navigate to another page
+  } else {
+    // Rerun spec
+  }
+  return next
+})
+```
+
+### Client Details:
+
+- [API Docs for useSubscription](https://formidable.com/open-source/urql/docs/api/urql/#usesubscription)
+- [Subscriptions overview](https://formidable.com/open-source/urql/docs/advanced/subscriptions/)
+
+ One thing to be aware of, is the subscription is only mounted/responded to when the containing component is mounted on the page.
+
+### Server Docs
+
+Subscriptions are implemented on the server as an `AsyncIterator`. This is handled for us by the [graphql-ws](https://github.com/enisdenjo/graphql-ws) package.
+
+To add a new Subscription field, add a new entry in the [`gql-Subscriptions`](../packages/graphql/src/schemaTypes/objectTypes/gql-Subscription.ts):
+
+```ts
+t.field('browserStatusChange', {
+  type: CurrentProject,
+  description: 'Status of the currently opened browser',
+  subscribe: (source, args, ctx) => ctx.emitter.subscribeTo('browserStatusChange'),
+  resolve: (source, args, ctx) => ({}),
+})
+```
+
+And then add the corresponding method in [DataEmitterActions](../packages/data-context/src/actions/DataEmitterActions.ts)
+
+
+```ts
+browserStatusChange () {
+  this._emit('browserStatusChange')
+}
+```
+
+- [API Docs](https://github.com/enisdenjo/graphql-ws/tree/master/docs)
+- [Transport layer protcol specification](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md)
+
+### Testing
+
+If you want to TDD the subscription being added & working directly in isolation, one recommended approach is to add one spec file per-subscription in the `/app/cypress/e2e/subscription` directory. This file can cover both the app & launchpad handling of a given subscription.
```

Comment:
I guess you might have thought about this already, but it feels weird to have launchpad tests written inside of `packages/app`?

As an aside, I wonder if eventually we want an entire package for testing out the subscriptions in isolation. For now, our test suite for GraphQL Subscriptions is tightly coupled to Cypress app's suite. 

An example use case would be we implement a public facing subscription for library authors to use, that we are not necessarily using ourselves. This is super long term vision, though.

Reply:
I was thinking of the `/app` tests as being a superset of the launchpad, since it has the servers mounted for both, so I don't see any problem with having them in there rather than needing to split into two files, it was easier to TDD having them co-located as opposed to jump between two projects to see if changes broke in one or another. 

> As an aside, I wonder if eventually we want an entire package for testing out the subscriptions in isolation.

That's sort of what I was looking to do with the `/subscriptions` folder, I don't know if we'd want to make this public facing for now, but if we go that route we probably will want to think about something like that